### PR TITLE
feat: Formalizes the config system for IOx, including tests

### PR DESCRIPTION
**Pull request by alamb**
Related to #599 

Editor's Note: I will admit to some partial yak shaving here, but I think the config documentation alone was worth it. I could keep going but I think this is a good enough place initially.

# Rationale
Rationale for a more structured config system is to ensure that the config values for IOx are clear, defined and documented in a single place, both for developers as well as users / operators. The tools in this PR are basically the ones I want to have when using server software (tell me what the current config values are, and what the possible values could be).

Prior to this PR, config is specified by several partially documented environment variables, and that documentation is in a file that can drift from the code.

This PR also aims to make developer's lives as simple as possible while enforcing documented config values: just follow the existing pattern rather than having to invent other config / documentation mechanisms. A little friction to adding new config values may also discourage a wild array of knobs -- and such wild arrays can be quite discouraging for users.

# User Visible Behavior
Configuration is (still) specified environment variables

For development or other testing purposes, the values for config settings can also be specified using a file (~/.influxdb_iox/confg) which uses ".dotenv" style

There is a `--ignore-config-file` command line argument that can be used to ignore the config file and take configuration only from the environment)

There are is a new command `config` which you can use to see the current config settings as well as all possible settings. Here are some examples:

```
# show current config values (that are non empty)
# (note I have set INFLUXDB_IOX_BIND_ADDR=127.0.0.1:11081 in my ~/.influxdb_iox/config file locally)
./target/debug/influxdb_iox config show
RUST_LOG=warn
INFLUXDB_IOX_BIND_ADDR=127.0.0.1:11081 # (default 127.0.0.1:8080)
INFLUXDB_IOX_GRPC_BIND_ADDR=127.0.0.1:8082
INFLUXDB_IOX_DB_DIR=/Users/alamb/.influxdb_iox/
```

```
# show configuration that comes only from the enviroment
./target/debug/influxdb_iox --ignore-config-file config show
WARNING: Ignoring config file "/Users/alamb/.influxdb_iox/config"
RUST_LOG=warn
INFLUXDB_IOX_BIND_ADDR=127.0.0.1:8080
INFLUXDB_IOX_GRPC_BIND_ADDR=127.0.0.1:8082
INFLUXDB_IOX_DB_DIR=/Users/alamb/.influxdb_iox/
```

Here is how to show all the possible config values
```
# show all available config options
./target/debug/influxdb_iox config help
Loading config from file and environment (file: "/Users/alamb/.influxdb_iox/config")
InfluxDB IOx Configuration:
-----------------
RUST_LOG: Rust logging level
  current value: warn
  default value: warn
  example value: debug,hyper::proto::h1=info

This controls the IOx server logging level, as described in https://crates.io/crates/env_logger. Levels for different modules can be specified as well. For example `debug,hyper::proto::h1=info` specifies debug logging for all modules except for the `hyper::proto::h1' module which will only display info level logging.
-----------------
OTEL_EXPORTER_JAEGER_AGENT_HOST: Open Telemetry Jaeger Host
  current value:
  example value: jaeger.influxdata.net

If set, Jaeger traces are emitted to this host using the OpenTelemetry tracer.

NOTE: The OpenTelemetry agent CAN ONLY be configured using environment variables. It CAN NOT be configured using the IOx config file at this time. Some useful variables:
 * OTEL_SERVICE_NAME: emitter service name (iox by default)
 * OTEL_EXPORTER_JAEGER_AGENT_HOST: hostname/address of the collector
 * OTEL_EXPORTER_JAEGER_AGENT_PORT: listening port of the collector.

The entire list of variables can be found in https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-environment-variables.md#jaeger-exporter
-----------------
INFLUXDB_IOX_ID: The identifier for the server
  current value:

The identifier for the server. Used for writing to object storage and asan identifier that is added to replicated writes, WAL segments and Chunks.Must be unique in a group of connected or semi-connected IOx servers.Must be a number that can be represented by a 32-bit unsigned integer.
-----------------
INFLUXDB_IOX_BIND_ADDR: HTTP bind address
  current value: 127.0.0.1:11081
  default value: 127.0.0.1:8080

The address on which IOx will serve HTTP API requests
-----------------
INFLUXDB_IOX_GRPC_BIND_ADDR: gRPC bind address
  current value: 127.0.0.1:8082
  default value: 127.0.0.1:8082

The address on which IOx will serve Storage gRPC API requests
-----------------
INFLUXDB_IOX_DB_DIR: Where to store files on disk:
  current value: /Users/alamb/.influxdb_iox/
  default value: /Users/alamb/.influxdb_iox/

The location InfluxDB IOx will use to store files locally
-----------------
INFLUXDB_IOX_GCP_BUCKET: The bucket name, if using Google Cloud Storage as an object store
  current value:
  example value: bucket_name

If using Google Cloud Storage for the object store, this item, as well as SERVICE_ACCOUNT must be set.
```

## Behavior Changes from main:
Configuration values are pulled from ~/.influxdb_iox/config if it exists rather than `.env` (but the format is the same)

Values in ~/.influxdb_iox/config that are not explicitly listed (e.g. OETL_JAEGER_AGENT_PORT) will be ignored. On main, any values in `.env` are  swizzled into IOx's environment (and thus can affect the open telemetry logging setup, for example)

# Code Changes
This PR adds:
1. A new crate for managing config (I know I love making crates....)
2. Port use of environment variable config (except for logging and object store) to use new code
2. Unit tests for all of the above
2. New command line options to show

# Follow on work

This PR doesn't really change config validation semantics (e.g. requiring a writer_id to be set) but it puts in place all the structure to make it easy to implement

Also, logging configuration still reads directly from the environment (via libraries for `env_logger` and `opentelemetry`) as I didn't want to change too much in this PR. If we want IOx explicitly configure opentelemetry we can plumb through the appropriate config values.

The object store crate still uses environment variables for the various object stores -- porting those config values over and validating that exactly one object store is configured is not done.

Add additional usability command  like `./influxdb_iox config generate` to create an initial config file (and set the writer id).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,25 +115,19 @@ InstalledDir: /Library/Developer/CommandLineTools/usr/bin
...
 
-For development purposes, the most relevant environment variables are the `INFLUXDB_IOX_DB_DIR` and
-`TEST_INFLUXDB_IOX_DB_DIR` variables that configure where files are stored on disk. The default
-values are shown in the comments in the example file; to change them, uncomment the relevant lines
-and change the values to the directories in which you'd like to store the files instead:
+You can see a list of all available configuration items using the `influxdb_iox config help` 
+command.
+
+Should you desire specifying config via a file, you can do using a `.env` formatted file in `$HOME/.influxdb_iox/config`. 
+Note that you can save the current config values by saving the otuput of
```
Comment (pauldix) [inline README.md:128]:
```suggestion
Note that you can save the current config values by saving the output of 
```

File: README.md [hunk 2]
```diff
@@ -115,25 +115,19 @@ InstalledDir: /Library/Developer/CommandLineTools/usr/bin
...
-then edit the newly-created `.env` file.
 
-For development purposes, the most relevant environment variables are the `INFLUXDB_IOX_DB_DIR` and
-`TEST_INFLUXDB_IOX_DB_DIR` variables that configure where files are stored on disk. The default
-values are shown in the comments in the example file; to change them, uncomment the relevant lines
-and change the values to the directories in which you'd like to store the files instead:
+You can see a list of all available configuration items using the `influxdb_iox config help` 
+command.
+
+Should you desire specifying config via a file, you can do using a `.env` formatted file in `$HOME/.influxdb_iox/config`.
```
Comment (pauldix) [inline README.md:127]:
```suggestion
Should you desire specifying config via a file, you can do so using a `.env` formatted file in `$HOME/.influxdb_iox/config`. 
```

File: README.md (hunk 1 above)
Comment (pauldix) [inline README.md:128]:
```suggestion
Note that you can save the current config values by saving the output of 
```

File: config/src/item.rs [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    fn name(&self) -> &'static str {
+        "INFLUXDB_IOX_ID"
+    }
+    fn short_description(&self) -> String {
+        "The identifier for the server".into()
+    }
+    // There is no default datbase ID (on purpose)
+    fn long_description(&self) -> Option<String> {
+        Some(
+            "The identifier for the server. Used for writing to object storage and as\
```
Comment (pauldix) [inline config/src/item.rs:188]:
```suggestion
            "The identifier for the server. Used for writing to object storage and as \
```

File: config/src/item.rs [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+             an identifier that is added to replicated writes, WAL segments and Chunks.\
```
Comment (pauldix) [inline config/src/item.rs:189]:
```suggestion
             an identifier that is added to replicated writes, WAL segments and Chunks. \
```

File: config/src/item.rs [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+             Must be unique in a group of connected or semi-connected IOx servers.\
```
Comment (pauldix) [inline config/src/item.rs:190]:
```suggestion
             Must be unique in a group of connected or semi-connected IOx servers. \
```

File: config/src/item.rs [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+}
+
+/// This value simply passed into the environment and used by the
+/// various loggign / tracing libraries. It has its own structure here
+/// for documentation purposes and so it can be loaded from config file.
+pub(crate) struct RustLog {}
+
+impl ConfigItem<Option<String>> for RustLog {
+    fn name(&self) -> &'static str {
+        "RUST_LOG"
```
Comment (pauldix) [inline config/src/item.rs:249]: Not for this PR, but it feels odd to me that we're using `RUST_LOG` as the log setting vs. `INFLUXDB_IOX_LOG_LEVEL`

File: config/src/item.rs [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+    }
+    fn short_description(&self) -> String {
+        "Rust logging level".into()
```
Comment (pauldix) [inline config/src/item.rs:252]: Is this a Rust thing or just a logging level set for IOx?

File: config/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+        example: String,
+        config_value: String,
+        message: String,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// The InfluxDB application configuration. This struct provides typed
+/// access to all IOx's configuration values.
```
Comment (pauldix) [inline config/src/lib.rs:83]:
```suggestion
/// access to all of IOx's configuration values.
```

**Review by pauldix — APPROVED** [review]
Just a couple of formatting/typo tweaks suggested. Overall looks great. I had a few questions for possible followup work.

File: config/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+    /// GCP object store bucekt
+    pub gcp_bucket: Option<String>,
+}
+
+impl Config {
+    /// Create the configuration object from a set of
+    /// name=value pairs
+    ///
+    /// ADD NEW CONFIG VALUES HERE
+    fn new_from_map(name_values: HashMap<String, String>) -> Result<Self> {
```
Comment (e-dard) [inline config/src/lib.rs:115]: Why does this return a `Result` that's always `Ok`?

**Review by e-dard — APPROVED** [review]
This looks good to me. Seems pretty similar to the way config is done in other Influx stuff. Only thing I wasn't sure of from the PR description is env overriding config file settings.

I would expect an env var to override any value in a configuration file; I would not expect to have to explicitly set or specify anything to make that override be used so to speak.

File: config/src/item.rs (hunk 6 above)
Comment (alamb) [reply config/src/item.rs:249]: I agree that this is odd -- this PR actually doesn't change how logging is configured, it merely documents it. So in my opinion that is progress, even though there is probably more to do. I filed https://github.com/influxdata/influxdb_iox/issues/609 to track this TODO

**Comment by alamb** [comment]
Filed https://github.com/influxdata/influxdb_iox/issues/610 to track the object store config improvements

**Comment by alamb** [comment]
> I would expect an env var to override any value in a configuration file; I would not expect to have to explicitly set or specify anything to make that override be used so to speak.

Great! This is the intended behavior (and what I believe the code does)

**Pull request merged**